### DISCUSSION TITLE
db model types

### DISCUSSION THREAD

**Discussion by jmaldon1** [mswjs/data#200] [Q&A] [answered]
I am not seeing a clear answer anywhere so I thought I'd just ask.

How do you add types for a model in typescript.

For example:

```typescript
export const db = factory({
  muscle: {
    id: primaryKey(Number),
    muscle: String,
  }
});
```

What should the type of this muscle array be?
```typescript
const muscles = [];
```

Do I need to create my own type:

```typescript
export type muscle = {
  id: number;
  muscle: string;
};
```

and then do this:

```typescript
const muscles: muscle[] = []
```

**Comment by kettanaito** [comment, marked as answer]
Hey, @jmaldon1. 

The types are inferred from the value getter you give to your model. This is done automatically by the library, and the result you get is already typed according to your model:

```ts
const db = factory({
  muscle: {
    id: primaryKey(String),
    muscle: string
  }
})

const muscles = db.muscle.findMany()
// Array<{ id: string, muscle: string }>
```

If you wish to _extract_ the type from the generated `db` client, you can do that from the `db[model].findFirst` return type:

```ts
type ExtractModelType<F extends Function> = ReturnType<F>

type Muscle = ExtractModelType<typeof db.muscle.findFirst>
// { id: strig, muscle: string }
```

I wouldn't advise doing this though. 

The library does not support accepting types as a generic to the `factory` just yet. You can follow and contribute to this in #36.

**Reply by jmaldon1** [reply]
So this is the usecase that I am having issues with:

```typescript
const muscles = [];

for (let i = 0; i < 10; i++) {
  const id = i + 1;

  muscles.push(db.muscle.create(createMuscleData(id)));
}

for (let i = 0; i < 10; i++) {
  const randomMuscles = [];  // Error: Variable 'randomMuscles' implicitly has type 'any[]' in some locations where its type cannot be determined.
  for (let i = 0; i < 3; i++) {
    const randomMuscle = muscles[Math.floor(Math.random() * muscles.length)];
    if (randomMuscles.includes(randomMuscle)) {  // Error: Variable 'randomMuscles' implicitly has an 'any[]' type.
      i--;
      continue;
    }
    randomMuscles.push(randomMuscle);
  }
}
```

It seems like I need to provide a type to `randomMuscles` before doing `.includes`

**Reply by kettanaito** [reply]
Okay, in this case you have to describe the data by yourself:

```ts
interface Muscle {
  id: string
  muscle: string
}

const muscles: Muscle[] = []
```

This is not strictly a library-related question.
